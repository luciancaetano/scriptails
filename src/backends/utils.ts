export function getProcessArgvBinaryIndex() {
    if (!!(process as any).versions.electron && !(process as any).defaultApp) return 0;
    return 1;
}

export function hideProcessArgvBinary(argv: string[]) {
    return argv.slice(getProcessArgvBinaryIndex() + 1);
}

export function getProcessArgvBinary(argv: string[]) {
    return argv[getProcessArgvBinaryIndex()];
}

export function isOption(str: string) {
    return str && (str.slice(0, 2) === '-' || str.slice(0, 1) === '-');
}
