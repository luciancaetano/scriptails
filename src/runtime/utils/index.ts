import * as chalk from 'chalk';
import * as program from 'commander';
import * as globLib from 'glob';

export * from './program';

const { log: consoleLog } = console;

const isSilent = () => program.opts().silent;

/**
 * Type in the terminal
 * @note Do nothing if the --silent flag is present
 * @param ...args:any[]
 */
export function log(...args: any[]) {
    if (isSilent()) return;
    consoleLog(...args);
}

/**
 * Write in the terminal each element of the array on a line
 * @note Do nothing if the --silent flag is present
 * @param args:string[]
 */
export function logLines(args: string[]) {
    if (isSilent()) return;
    args.forEach((line) => consoleLog(line));
}

/**
 * Write in the terminal with a label as a prefix, this label is stylized and formatted according to its context.
 * @note Do nothing if the --silent flag is present
 * @param text string
 * @param label 'error'|'info'|'warning'|'success'
 */
export function logWithLabel(text: string, label: 'error' | 'info' | 'warning' | 'success') {
    if (isSilent()) return;

    let currLabel = '';

    switch (label) {
    case 'error':
        currLabel = chalk.bgRed.white.bold(' ERROR ');
        break;

    case 'info':
        currLabel = chalk.bgBlue.white.bold(' INFO ');
        break;

    case 'warning':
        currLabel = chalk.bgYellow.white.bold(' WARNIGN ');
        break;

    case 'success':
        currLabel = chalk.bgGreen.white.bold(' SUCCESS ');
        break;
    default:
        currLabel = '';
        break;
    }

    consoleLog(`${currLabel} ${text}`);
}

/**
 * Check if proccess is runing on expected platform
 * @param expected expected platform 'aix' | 'darwin' | 'freebsd' | 'linux' | 'openbsd' | 'sunos' | 'win32
 */
export function isPlatform(expected: 'aix' | 'darwin' | 'freebsd' | 'linux' | 'openbsd' | 'sunos' | 'win32') {
    return process.platform === expected;
}

/**
 * Create an promisified glob see more in https://github.com/isaacs/node-glob
 */
export const glob = (pattern: string, options: globLib.IOptions) => new Promise<string[]>((resolve, reject) => {
    globLib(pattern, options, (err, files) => (err === null ? resolve(files) : reject(err)));
});
