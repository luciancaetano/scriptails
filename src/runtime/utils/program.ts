import {
    filter, first, get, isError, map,
} from 'lodash';
import * as chalk from 'chalk';
import { program } from 'commander';
import ScriptContext from '../../ScriptContext';
import MixedType from '../core/castable/MixedType';

const { log: consoleLog } = console;

/**
 * Print an error and exit
 *
 * @param {string|Error} message
 * @param {boolean} label
 */
export function exitError(message: string | Error, label = true) {
    consoleLog(`${label ? `${chalk.bgRed.white.bold(' ERROR ')}${chalk.red.bold(':')} ` : ''}${chalk.red.bold(isError(message) ? message.message : message)}`);
    process.exit(1);
}

/**
 * Get all options/flags in program
 * @returns {Object}
 */
export function getOptions() {
    const commandName = ScriptContext.getInstance().getCurrentCommandName();

    return first(filter(map(program.commands, (command) => {
        if (command.name() === commandName) {
            return command.opts();
        }
        return null;
    }), (item) => !!item)) || {};
}

/**
 * Read an option/flag in program
 * @param {string} name
 * @param {string|boolean|null} defaultValue
 */
export function getOption(name: string, defaultValue: string | boolean | null = null) {
    return new MixedType(get(getOptions(), name, defaultValue) || defaultValue);
}
