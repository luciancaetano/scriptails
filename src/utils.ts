export function isPresent(value: any) {
    return value !== null && value !== undefined && String(value).trim().length > 0;
}

export function getOptionName(flags: string[]): string {
    // Check for flags
    flags.forEach((flag) => {
        if (!(flag.slice(0, 1) === '-' || flag.slice(0, 2) !== '--')) {
            throw new Error(`Invalid flag "${flag}", flags must start with "-" for short flags or "--" for long flags.`);
        } else if (flag.includes(' ')) {
            throw new Error(`Invalid flag "${flag}", flags must not contains spaces.`);
        }
    });

    const firstLongFlag = flags.find((flag) => flag.slice(0, 2) === '--');

    if (firstLongFlag && firstLongFlag.trim().length > 0) {
        return firstLongFlag.slice(2, firstLongFlag.length);
    }

    const firstShortFlag = flags.find((flag) => flag.slice(0, 1) === '-');

    if (firstShortFlag && firstShortFlag.trim().length > 0) {
        return firstShortFlag.slice(1, firstShortFlag.length);
    }

    console.info(`firstLongFlag: ${firstLongFlag} \nfirstShortFlag: ${firstShortFlag}\n `);

    throw new Error(`Invalid flags ['${flags.join('\', \'')}'], flags must start with "-" for short flags or "--" for long flags.`);
}
