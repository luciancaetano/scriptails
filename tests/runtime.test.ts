import { cli } from './utils';

describe('Runtime lib testing', () => {
    it('should run a cli with help arg', async () => {
        const result = await cli(['--help'], '.');

        expect(result.code).toBe(0);
        expect(result.stdout).toMatchSnapshot();
    });

    it('should match cli version', async () => {
        const result = await cli(['--version'], '.');
        expect(result.code).toBe(0);
    });

    it('should run utils.exec', async () => {
        const result = await cli([], '.');
        expect(result.code).toBe(0);
    });

    it('should run utils.execFile', async () => {
        const result = await cli([], '.');
        expect(result.code).toBe(0);
    });

    it('should logLines', async () => {
        const result = await cli([], '.');
        expect(result.code).toBe(0);
    });

    it('should logWithLabel', async () => {
        const result = await cli([], '.');
        expect(result.code).toBe(0);
    });

    it('should log a text', async () => {
        const result = await cli([], '.');
        expect(result.code).toBe(0);
    });

    it('should not logLines when flag silent is set', async () => {
        const result = await cli([], '.');
        expect(result.code).toBe(0);
    });

    it('should not logWithLabel when flag silent is set', async () => {
        const result = await cli([], '.');
        expect(result.code).toBe(0);
    });

    it('should not log a text when flag silent is set', async () => {
        const result = await cli([], '.');
        expect(result.code).toBe(0);
    });
});
