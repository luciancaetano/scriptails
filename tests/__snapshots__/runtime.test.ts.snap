// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Runtime lib testing should run a cli with help arg 1`] = `
"
> scriptails@2.0.1 test:cli D:\\\\projects\\\\scriptmizer
> ts-node ./tests/mocked-cli.ts \\"--help\\"

Usage: mocked-cli [options] [command]

Options:
  --silent        Disable log and debug information, including children
                  proccess stdout (default: false)
  -V, --version   output the version number
  -h, --help      display help for command

Commands:
  list-files      it should list files
  help [command]  display help for command
"
`;
